Implement BooksController#Show and BooksController#Index

cquiro opened the pull request:
## Summary

* Add and configure gems versionist, active_model_serializers and wor-paginate
* Create BooksController and BookSerializer
* Create endpoint for books#show
* Create endpoint for books#index (including pagination)
* Require authentication for all actions except the ones that are performed by the devise_controller

## Trello Card

https://trello.com/c/K1gUlUqv/23-crear-rutas-index-y-show-para-book

cquiro committed:
Create show endpoint for books

- Add gem versionist to Gemfile.

- Add gem active_model_serializers to Gemfile.

- Bundle install.

- Create serializer for book.

- Create books controller with show action.

- Create route for api/v1/books/:id

cquiro committed:
Return the image url directly.

cquiro committed:
Create initial code for books endpoint

- Add route for api/v1/books

- Add index action to BooksController and return Book.all

- No pagination or filters implemented yet

cquiro committed:
Paginate index of Api::V1::BooksController

- Add wor-paginate gem and bundle install

- Run rails generate wor:paginate:install to create
configuration file

- Include Wor::Paginate in BooksController and use it to
paginate index method

cquiro committed:
Authenticate user before_action unless it's a devise_controller action.

facucompe commented on app/serializers/api/v1/book_serializer.rb:
```diff
@@ -0,0 +1,8 @@
+module Api
```
No need to use `api/v1` folders

facucompe commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,25 @@
+module Api
+  module V1
+    class BooksController < ApplicationController
+      include Wor::Paginate
```
Move this to application controller

facucompe commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,25 @@
+      private
+
+      def book
+        @book ||= Book.find(params[:id])
```
No need to use instance variable or memoization.

facucompe replied:
Also use a separate method to permit params.
```
def book_params
  params.permit(:id)
end
```

facucompe commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,25 @@
+      end
+
+      def books
+        @books ||= Book.all
```
Same here. No need to use an instance variable and memoization

facucompe commented on app/controllers/application_controller.rb:
```diff
@@ -3,6 +3,7 @@ class ApplicationController < ActionController::Base
 
   include DeviseTokenAuth::Concerns::SetUserByToken
 
+  before_action :authenticate_user!, unless: :devise_controller?
```
We usually use an `api_controller` that inherits from `application_controller` where we put the authentication.

facucompe commented:
Try using **spinal-case** for branch names instead of **snake_case**

cquiro committed:
Make changes as per code review

- Put book_serializer.rb into app/serializers

- Create api_controller that inherits from application_controller
and put there api specific code

- Add book_params method to books_controller.rb

- Stop using instance variables and memoization in books_controller.rb

lmasello reviewed (approved):
Just what facu asked and then it's good to go 👍

cquiro committed:
Update namespaces of the code

- Remove Api::V1 namespace for BookSerializer

- Move 'protect_from_forgery with: :null_session' back to
application_controller.rb

- Move ApiController inside Api::V1 namespace

cquiro committed:
Merge branch 'master' into show_index_books

facucompe commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -0,0 +1,27 @@
+  module V1
+    class BooksController < ApiController
+      def index
+        render_paginated books, each_serializer: Api::V1::BookSerializer, status: :ok
```
Watch out, you are not serializing. It should just be `BookSerializer`

cquiro replied:
@facucompe Done!

cquiro committed:
Replace Api::V1::BookSerializer with BookSerializer.
